File: compiler/src/macro_registry.py
from dataclasses import dataclass
from io import StringIO
from typing import TYPE_CHECKING, Callable, Union, TypeVar, Protocol, cast, Any

from node import Node
from strutil import IndentedStringIO
from logger import default_logger

if TYPE_CHECKING:
    from processor import MacroProcessingStep
    from macrocosm import Macrocosm

# TODO - this shouldn't be here, probably...
@dataclass(kw_only=True)
class MacroContext:
    statement_out: IndentedStringIO | StringIO
    expression_out: IndentedStringIO | StringIO
    node: Node

    compiler: "Macrocosm"
    current_step: "MacroProcessingStep | None" = None

class Macro_preprocess_provider(Protocol):
    def preprocess(self, ctx: MacroContext): ...
class Macro_typecheck_provider(Protocol):
    def typecheck(self, ctx: MacroContext) -> str | None: ...
class Macro_emission_provider(Protocol):
    def emission(self, ctx: MacroContext): ...

class Macro_code_linking_provider(Protocol):
    def code_linking(self, ctx: MacroContext): ...

Macro_provider = \
    Macro_preprocess_provider | \
    Macro_typecheck_provider | \
    Macro_emission_provider

class Macro(Protocol):
    def __call__(self, ctx: MacroContext) -> str: ...

F = TypeVar("F", bound=Callable[..., Any])

class MacroRegistry:
    def __init__(self) -> None:
        self._registry: dict[str, Macro] = {}

    def add_fn(self, m: Macro | None, *names: str):
        if m is None:
            return
        for name in names:
            self._registry[name] = m

    def get(self, name: str) -> Macro:
        try:
            return self._registry[name]
        except KeyError:
            default_logger.macro(f"ERROR: unknown macro '{name}'")
            raise ValueError(f"Unknown macro: {name}")

    def all(self) -> dict[str, Macro]:
        return dict(self._registry)  # defensive copy

File: compiler/src_ts/core/macro_registry.ts
// core/macro_registry.ts

import { IndentedStringIO } from "../utils/strutil.ts";
import { default_logger } from "../utils/logger.ts";
import type { Node } from "./node.ts";
import type { Macrocosm } from "./macrocosm.ts";
import { Constructor } from "../utils/utils.ts";
; 

export type OutIO = IndentedStringIO | { 
  write: (s: string) => void,
  gets_to_end: () => string
};

// --------------------------------------
// MacroContext
// --------------------------------------

export interface Macro_context<T = unknown> {
  node: Node;
  compiler: Macrocosm;
  clone_with(opts: {
    node: Node
  }): Macro_context;
  apply(): T;
}

export const REGISTER_MACRO_PROVIDERS = Symbol("REGISTER_MACRO_PROVIDERS");

export interface Macro_provider {
  [REGISTER_MACRO_PROVIDERS](via: Register_macro_providers): void;
}

export type Register_macro_providers = <T extends Macro_context>(type: Constructor<T>, macro: string, fn: Ctx_proc<T>) => void;


// --------------------------------------
// MacroRegistry
// --------------------------------------

export type Ctx_proc<T extends Macro_context> = (ctx: T) => ReturnType<T["apply"]>

export class Macro_registry<T extends Macro_context> {
  private registry: Record<string, Ctx_proc<T>>;

  constructor() {
    this.registry = {};
  }

  add_fn(fn: Ctx_proc<T> | null, ...names: string[]): void {
    if (fn === null) {
      return;
    }
    for (const name of names) {
      this.registry[name] = fn;
    }
  }

  get(name: string): Ctx_proc<T> {
    const fn = this.registry[name];
    if (!fn) {
      throw new Error(`Unknown macro: ${name}`);
    }
    return fn;
  }

  all(): Record<string, Ctx_proc<T>> {
    return { ...this.registry };
  }
}
